rubin_scheduler/utils/consdb.py

```diff
@@ -0,0 +1,1057 @@
+            # environment without it.
+            from lsst.summit.utils import ConsDbClient  # type: ignore
+
+            consdb = ConsDbClient(url)
+            query_results: pd.DataFrame = consdb.query(query).to_pandas()
+        case _:
+            raise ValueError(f"Unrecongined url scheme {url_scheme} in consdb url {url}")
+
+    return query_results
+
+
+# Different instruments (and different numbers of snaps per visit) do or
+# will have differenc schema in the consdb, so make a base class with
+# expected common functionality, and use derived classes for
+# instrument-specific elements.
+
+# Separate different pieces of metadata (columns in a table of visits) into
+# different methods to make it easy to individually override them by
+# instrument.
+
+
+# cached_properties are a way to compute values when and only when needed
+# but these won't update if the date is changed. In my expected use, that's
+# fine; the user should just instantiate another ConsDBVisits instance if
+# they want a different database or date. But, what if someone does?
+# To avoid this, make it a frozen dataclass, which prevents members from
+# being updated.
+@dataclass(frozen=True)
+class ConsDBVisits(ABC):
+    """Visits returned from a query to the consdb.
+
+    Parameters
+    ----------
+    day_obs : `str`
+        The date for which to query the database.
+    url : `str`
+        The connection string from the database.
+    """
+
+    day_obs: str | int
+    url: str = "postgresql://usdf@usdf-summitdb.slac.stanford.edu:5432/exposurelog"
+
+    @property
+    @abstractmethod
+    def instrument(self) -> str:
+        """The instrument.
+
+        Returns
+        -------
+        instrument : `str`
+            The instrument.
+        """
+        # Subclasses should return a string constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def num_exposures(self) -> int:
+        """The number of snaps per visit.
+
+        Returns
+        -------
+        num_exposures : `int`
+            The number of snaps per visit.
+        """
+        # Subclasses should return a string constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def telescope(self) -> str:
+        """The telescope.
+
+        Returns
+        -------
+        telescope : `str`
+            The telescope.
+        """
+        # Subclasses should return a constant specifying the relevant
+        # instrument.
+        # Specified as an property rather than an attribute so it can be
+        # abstract.
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def gain(self) -> pd.Series:
```

Comment:
Gain is not available in the consdb. If it becomes available, it will be easy to update the class.
For some of what we need, this needs to work properly before DM processes the data (but after quicklook). We may want to have separate subclasses for when the "real" DM outputs are available, and when only quicklook ones are.

Reply:
Prompt processing should also be very fast, so ought to be also available. I think there is currently a question running around about how to keep parallel but not identical outputs between RA and PP.